tt_metal/impl/program/program_pool.cpp

```diff
@@ -0,0 +1,25 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "tt_metal/impl/program/program_pool.hpp"
+
+namespace tt::tt_metal {
+
+ProgramHandle ProgramPool::create_program() {
+    std::unique_lock lock(mutex_);
+    ProgramKey key = programs.emplace();
+    return ProgramHandle{key.raw()};
+}
+
+void ProgramPool::release_program(ProgramHandle handle) {
+    std::unique_lock lock(mutex_);
```

Comment:
If we have multiple device threads working with programs they will be waiting each other in this singleton?
Any ideas how to avoid so many locks and singleton?

Reply:
Yes, it would require the underlying slotmap to become concurrency-aware and lock-free. It's definitely something I want to address, once we know the lock-based implementation is correct.

Reply:
Lockfree  doesn't mean noone is waiting. It might be even slower than mutexes in this case.
It will still require unnecessary synchronizations between unrelated things.
There is always obvious owner of the program. For example Device or MeshDevice. Program pool should live here.